cupy/manipulation/basic.py

```diff
@@ -69,3 +70,69 @@ def _can_memcpy(dst, src):
     f_contiguous = dst.flags.f_contiguous and src.flags.f_contiguous
     return (c_contiguous or f_contiguous) and dst.dtype == src.dtype and \
         dst.size == src.size
+
+
+_putmask_kernel = core._kernel.ElementwiseKernel(
+    'Q mask, raw S values, uint64 len_vals', 'T out',
+    '''
+    if (mask) out = (T) values[i % len_vals];
+    ''',
+    'putmask_kernel'
+)
+
+
+def putmask(a, mask, values):
+    """
+    Changes elements of an array inplace, based on conditional mask and
+    input values.
+
+    Sets ``a.flat[n] = values[n]`` for each n where ``mask.flat[n]==True``.
+    If `values` is not the same size as `a` and `mask` then it will repeat.
+
+    Args
+        a (cupy.ndarray): Target array.
+        mask (cupy.ndarray): Boolean mask array. It has to be
+            the same shape as `a`.
+        values (cupy.ndarray or scalar): Values to put into `a` where `mask`
+            is True. If `values` is smaller than `a`, then it will be
+            repeated.
+
+    Examples
+    --------
+    >>> x = cupy.arange(6).reshape(2, 3)
+    >>> cupy.putmask(x, x>2, x**2)
+    >>> x
+    array([[ 0,  1,  2],
+           [ 9, 16, 25]])
+
+    If `values` is smaller than `a` it is repeated:
+
+    >>> x = cupy.arange(6)
+    >>> cupy.putmask(x, x>2, [-33, -44])
+    >>> x
+    array([  0,  1,  2, -44, -33, -44])
+
+    .. seealso:: :func:`numpy.putmask`
+
+    """
+    # TODO: support different dtypes for a and values without typecasting
+
+    if not a.shape == mask.shape:
+        raise ValueError('mask and data must be the same size')
+
+    if cupy.isscalar(values):
+        a[mask] = values
+
+    else:
+        if not numpy.can_cast(values.dtype, a.dtype):
+            raise TypeError("Cannot cast array data from dtype('float64') to "
+                            "dtype('int64') according to the rule 'safe'")
+
+        if a.shape == values.shape:
+            a[mask] = values[mask]
+
+        else:
+            if values.ndim > 1:
+                values = values.flatten()
```

Comment:
For the performance.
```suggestion
            values = values.ravel()
```